File: clojure/robot-name/src/robot_name.clj
(ns robot-name)

(def used-names (ref #{}))

(defn- rand-letter []
  (char (+ (int \A) (rand 26))))

(defn- unique-name []
  (let [name (str (rand-letter) (rand-letter) (rand-int 1000000))]
    (if-not (some #{name} @used-names)
      (dosync
        (ref-set used-names (conj @used-names name))
        name)
      (recur))))

(defn robot [] (ref (unique-name)))

(defn reset-name [robot]
  (let [new-name (unique-name)]
    (dosync
     (ref-set used-names (remove #(= name %) @used-names))
     (ref-set robot new-name))))

(defn robot-name [robot] @robot)

File: go/robot-name/robot_name.go
package robotname

import (
	"fmt"
	"math/rand"
)

type Robot struct {
	assignedName string
}

// Using a map means name lookups are O(1)
var names = map[string]bool{}

func (r *Robot) Name() string {
	if r.assignedName != "" {
		return r.assignedName
	}
	r.assignedName = genName()
	for names[r.assignedName] {
		r.assignedName = genName()
	}
	names[r.assignedName] = true
	return r.assignedName
}

func (r *Robot) Reset() {
	r.assignedName = ""
}

func genName() string {
	return fmt.Sprintf("%c%c%03d", randLetter(), randLetter(), rand.Int31n(1000))
}

func randLetter() byte {
	return byte(int32('A') + rand.Int31n(26))
}
